src/local_storage.rs

```diff
@@ -73,35 +76,33 @@ impl LocalStorage {
             .map(|any| {
                 any.downcast_ref::<T::Value>().ok_or_else(|| {
                     error!(
-                        "Could not downcast storage entry. T: {:?}, participant_id: {}",
-                        TypeId::of::<T>(),
+                        "Could not downcast storage entry. Type: {:?}, participant_id: {}",
+                        std::any::type_name::<T::Value>(),
                         participant_id
                     );
                     InternalError::InternalInvariantFailed
                 })
             })
-            .unwrap_or(Err(InternalError::StorageItemNotFound))
+            .unwrap_or_else(|| {
+                error!(
+                    "Could not retrieve storage entry. Type: {:?}, participant_id: {}",
+                    std::any::type_name::<T::Value>(),
+                    participant_id
+                );
+                Err(InternalError::InternalInvariantFailed)
+            })
     }
 
     /// Retrieves a mutable reference to a value via its [`TypeTag`]
     /// and [`ParticipantIdentifier`].
     pub(crate) fn retrieve_mut<T: TypeTag>(
         &mut self,
         participant_id: ParticipantIdentifier,
-    ) -> Result<&mut T::Value> {
-        self.storage
-            .get_mut(&(participant_id, TypeId::of::<T>()))
-            .map(|any| {
-                any.downcast_mut::<T::Value>().ok_or_else(|| {
-                    error!(
-                        "Could not downcast storage entry. T: {:?}, participant_id: {}",
-                        TypeId::of::<T>(),
-                        participant_id
-                    );
-                    InternalError::InternalInvariantFailed
-                })
-            })
-            .unwrap_or(Err(InternalError::StorageItemNotFound))
+    ) -> Option<&mut T::Value> {
+        match self.storage.get_mut(&(participant_id, TypeId::of::<T>())) {
+            Some(any) => any.downcast_mut::<T::Value>(),
+            None => None,
+        }
```

Comment:
suggestion: Another option here is to use `map` instead of matching. Either way is fine though!

`self.storage.get_mut(...).map(|any| any.downcast_mut::<T::Value>())`

Reply:
I think it would have to be
`.map_or(None, |any| any.downcast_mut::<T::Value>())`
which seems slightly less clear to me at least